cascadenet/network/layers/kspace_pytorch.py

```diff
@@ -0,0 +1,440 @@
+    -----------------------------
+    incomings: two 5d tensors, [kspace_data, mask], each of shape (n, 2, nx, ny, nt)
+
+    data_shape: shape of the incoming tensors: (n, 2, nx, ny, nt) (This is for convenience)
+
+    frame_dist: a list of distances of neighbours to sample for each averaging channel
+        if frame_dist=[1], samples from [-1, 1] for each temporal frames
+        if frame_dist=[3, 5], samples from [-3,-2,...,0,1,...,3] for one,
+                                           [-5,-4,...,0,1,...,5] for the second one
+
+    divide_by_n: bool - Decides how averaging will be done.
+        True => divide by number of neighbours (=#2*frame_dist+1)
+        False => divide by number of nonzero contributions
+
+    clipped: bool - By default the layer assumes periodic boundary condition along temporal axis.
+        True => Averaging will be clipped at the boundary, no circular references.
+        False => Averages with circular referencing (i.e. at t=0, gets contribution from t=nt-1, so on).
+
+    Returns
+    ------------------------------
+    output: 5d tensor, missing lines of k-space are filled using neighbouring frames.
+            shape becomes (n, (len(frame_dist))* 2, nx, ny, nt)
+    '''
+
+    def __init__(self, frame_dist, divide_by_n=False, clipped=True, norm='ortho'):
+        super(AveragingInKspace, self).__init__()
+        self.normalized = norm == 'ortho'
+        # self.fft2_fun = Fft2d(norm)
+        # self.ifft2_fun = Ifft2d(norm)
+        self.frame_dist = frame_dist
+        self.divide_by_n = divide_by_n
+        self.kavg = KspaceFillNeighbourLayer(frame_dist, divide_by_n, clipped)
+
+    def forward(self, *input, **kwargs):
+        return self.perform(*input)
+
+    def perform(self, x, mask):
+        """
+        x    - input in image space, shape (n, 2, nx, ny, nt)
+        mask - corresponding nonzero location
+        """
+        mask = mask.permute(0, 1, 4, 2, 3)
+
+        x = x.permute(0, 4, 2, 3, 1) # put t to front, in convenience for fft
+        k = torch.fft(x, 2, normalized=self.normalized)
+        k = k.permute(0, 4, 1, 2, 3) # then put ri to the front, then t
+
+        # data sharing
+        # nc is the numpy of copies of kspace, specified by frame_dist
+        out = self.kavg.perform(k, mask)
+        # after datasharing, it is nb, nc, 2, nt, nx, ny
+
+        nb, nc, _, nt, nx, ny = out.shape # , jo's version
+
+        # out.shape: [nb, 2*len(frame_dist), nt, nx, ny]
+        # we then detatch confused real/img channel and replica kspace channel due to datasharing (nc)
+        out = out.permute(0,1,3,4,5,2) # jo version, split ri and nc, put ri to the back for ifft
+        x_res = torch.ifft(out, 2, normalized=self.normalized)
+
+
+        # now nb, nc, nt, nx, ny, ri, put ri to channel position, and after nc (i.e. within each nc)
+        x_res = x_res.permute(0,1,5,3,4,2).reshape(nb, nc*2, nx,ny, nt)# jo version
+
+        return x_res
+
+# class AveragingInKspace(nn.Module):
```

Comment:
Delete this. Can you also make sure the functions not needed are not added? You can keep data sharing aspects